### PR TITLE
Saml provider

### PR DESCRIPTION
**Pull request by mrajashree**
Changes introduced because of using updated saml library
1. The 1.6 code would accept and pass private key and certificate in string format, and the saml library would convert it to key to use. Now it accepts rsa.PrivateKey format, so in the code we need to parse the string to get the rsa.PrivateKey ([starting here](https://github.com/rancher/rancher/pull/12320/files#diff-5a43c1b3c69ff7b6afe233e552d379d7R94))
2. The saml lib would return assertion in form of request headers, instead now it returns them in request context.
3. Some fields in the library, such as Metadata being renamed to EntityDescriptor along with some field changes in it

https://github.com/rancher/rancher/issues/11169

### CODE REVIEW

**Comment by dnauck** [comment]
Will this PR be merged in the 2.0 release? I'm interested to use SAML SSO with ADFS like all our other services (Office 365, Gitlab, etc.).

**Comment by jammm** [comment]
Supporting more SAML providers would be great! For example, GCP combined with Google cloud identity would be a great combination for those who primarily use GKE clusters. (I'm one of those ;))

File: pkg/auth/providers/saml/saml_provider.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+func (s *Provider) AuthenticateUser(input interface{}) (v3.Principal, []v3.Principal, map[string]string, error) {
+	login, ok := input.(*v3public.CodeBasedLogin)
+	if !ok {
+		return v3.Principal{}, nil, nil, errors.New("SAML: unexpected input type")
+	}
+
+	return s.LoginUser(login, nil, false)
+}
+
+func (s *Provider) LoginUser(samlCredential *v3public.CodeBasedLogin, config *v3.SamlConfig, test bool) (v3.Principal, []v3.Principal, map[string]string, error) {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: i know you just copied from github provider, but this function doesnt need to be exported

File: pkg/auth/providers/saml/saml_client.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	"time"
+
+	"github.com/crewjam/saml"
+	"github.com/crewjam/saml/samlsp"
+	"github.com/gorilla/mux"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	log "github.com/sirupsen/logrus"
+)
+
+//SPClient implements a client for shibboleth and the saml library
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: please replace all references to shibboleth or shib with generic "saml"

File: pkg/auth/providers/saml/saml_provider.go [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+			userPrincipal = s.toPrincipal(userType, a, nil)
+			userPrincipal.Me = true
+		} else {
+			groupPrincipal := s.toPrincipal(groupType, a, nil)
+			groupPrincipal.MemberOf = true
+			groupPrincipals = append(groupPrincipals, groupPrincipal)
+		}
+	}
+
+	testAllowedPrincipals := config.AllowedPrincipalIDs
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: I know this was just copied from another provider, but the name `testAllowedPrincipals` is confusing. The word "test" doesnt make sense to me. Can you change it to just `allowedPrincipals`

File: pkg/auth/providers/saml/saml_provider.go (hunk 1 above)
Comment (cjellick) [reply pkg/auth/providers/saml/saml_provider.go]: In general, can you review all your functions and ensure you are only exporting ones that actually need to be called from a different package?

File: tests/core/test_auth_configs.py [hunk 4]
```diff
@@ -12,20 +12,28 @@ def test_auth_configs(mc):
...
             ad = c
+        elif c.type == "pingSamlConfig":
+            saml = c
+        elif c.type == "authConfig":
+            auth = c
 
-    for x in [gh, local, ad]:
+    for x in [gh, local, ad, saml, auth]:
         assert x is not None
+        print("x")
```
Comment (cjellick) [inline tests/core/test_auth_configs.py]: your probably just debugging your test, but make sure you remove these lines

File: pkg/auth/tokens/server.go [hunk 5]
```diff
@@ -28,6 +28,7 @@ type tokenAPIServer struct {
 	tokensClient v3.TokenInterface
 	userIndexer  cache.Indexer
 	tokenIndexer cache.Indexer
+	//authConfigIndexer cache.Indexer
```
Comment (cjellick) [inline pkg/auth/tokens/server.go]: remove this line

File: app/authconfig_data.go [hunk 6]
```diff
@@ -20,7 +21,15 @@ func addAuthConfigs(management *config.ManagementContext) error {
...
-	return addAuthConfig(localprovider.Name, client.LocalConfigType, true, management)
+	if err := addAuthConfig(ping.Name, client.PingSamlConfigType, false, management); err != nil {
+		return err
+	}
+
+	if err := addAuthConfig(localprovider.Name, client.LocalConfigType, true, management); err != nil {
+		return err
+	}
+
+	return addAuthConfig("authconfig", client.AuthConfigType, false, management)
```
Comment (cjellick) [inline app/authconfig_data.go]: why did you add this line? I'd expect this file to _only_ be adding ping, nothing else.

File: pkg/auth/providers/saml/saml_actions.go [hunk 7]
```diff
@@ -50,8 +48,8 @@ func (s *Provider) configureTest(actionName string, action *types.Action, reques
 			fmt.Sprintf("SAML: Failed to parse body: %v", err))
 	}
 
-	redirectURL := formSamlRedirectURL(samlConfig)
-	err := s.SamlClient.InitializeSPClient(samlConfig)
+	redirectURL := s.formSamlRedirectURL(samlConfig)
+	err := s.InitializeSPClient(samlConfig, PingName)
```
Comment (mrajashree) [inline pkg/auth/providers/saml/saml_actions.go]: I need to change this to s.Name

File: pkg/auth/providers/saml/saml_provider.go [hunk 8]
```diff
@@ -30,7 +36,45 @@ type Provider struct {
...
 
-var Samlp *Provider
+func Configure(ctx context.Context, mgmtCtx *config.ScaledContext, userMGR user.Manager, name string) common.AuthProvider {
+	samlClient := &SPClient{}
+	pingSamlp := &Provider{
+		Ctx:         ctx,
+		AuthConfigs: mgmtCtx.Management.AuthConfigs(""),
+		UserMGR:     userMGR,
+		SamlClient:  samlClient,
+		Name:        PingName,
```
Comment (mrajashree) [inline pkg/auth/providers/saml/saml_provider.go]: Change this to name

File: pkg/auth/providers/saml/saml_provider.go [hunk 9]
```diff
@@ -30,7 +36,45 @@ type Provider struct {
...
+}
+
+func (s *Provider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = s.actionHandler
+	schema.Formatter = s.formatter
+}
+
+func (s *Provider) TransformToAuthProvider(authConfig map[string]interface{}) map[string]interface{} {
+	p := common.TransformToAuthProvider(authConfig)
+	if s.Name == "ping" {
```
Comment (mrajashree) [inline pkg/auth/providers/saml/saml_provider.go]: Use the constant

File: pkg/auth/providers/saml/saml_client.go [hunk 10]
```diff
@@ -0,0 +1,277 @@
...
+
+	"github.com/crewjam/saml"
+	"github.com/crewjam/saml/samlsp"
+	"github.com/gorilla/mux"
+	"github.com/rancher/types/apis/management.cattle.io/v3"
+	log "github.com/sirupsen/logrus"
+)
+
+//SPClient implements a client for the saml library
+type SPClient struct {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: we talked about changing sp to something else. it is not obvious what it means

File: pkg/api/controllers/samlconfig/samlconfig.go [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+		err := json.Unmarshal([]byte(samlConfStr), samlConfig)
+		if err != nil {
+			return fmt.Errorf("SAML controller: %v", err)
+		}
+	} else {
+		return nil
+	}
+
+	samlProvider := saml.Provider{}
+	return samlProvider.InitializeSPClient(samlConfig, config.Name)
```
Comment (cjellick) [inline pkg/api/controllers/samlconfig/samlconfig.go]: This function makes use of state on the samlProvider struct
Specifically:
```
	samlURL += s.Name
```
and
```
		Root.Get("PingLogin").Handler(samlspInstance.RequireAccount(http.HandlerFunc(s.HandleSamlPost)))
```

So, when you just create a new struct here and call the method, you are not going to get the desired result. The function `InitializeSPClient` should not be attached to any struct. It should be a stand alone function. It is going to have to "find" the correct provider from a map or something based on a name passed to it. Similar to how these functions work:
https://github.com/rancher/rancher/blob/33d900d485929bffbc907011ae47043de1cbf13b/pkg/auth/providers/providers.go#L27-L38

In fact, you should just use one of those. The only catch is that it returns a Provider interfact that will not a function exposed for configuring saml. So, you should create an interface specific to saml with the necessary reconfigure function on it and type switch it to your interface.

File: pkg/api/controllers/samlconfig/samlconfig.go [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+	apiContext.Management.AuthConfigs("").AddHandler("authConfigController", sync)
+}
+
+func sync(key string, config *v3.AuthConfig) error {
+	samlConfig := &v3.SamlConfig{}
+	if key == "" || config == nil {
+		return nil
+	}
+
+	if samlConfStr, ok := config.Annotations["tempSamlConfigbyte"]; ok {
```
Comment (cjellick) [inline pkg/api/controllers/samlconfig/samlconfig.go]: I thought we were going to drop the annotation and just do this based off the config itself. I think I understand your problem now though: the type AuthCOnfig is hiding the saml specific fields. I think you can use this paradigm to get and update the specific type:
get: https://github.com/rancher/rancher/blob/4d560ca5a8e5391a62fda1e0a00e334fb6153c73/pkg/auth/providers/github/github_provider.go#L66-L69

update: 
https://github.com/rancher/rancher/blob/4d560ca5a8e5391a62fda1e0a00e334fb6153c73/pkg/auth/providers/github/github_provider.go#L92

File: server/server.go [hunk 13]
```diff
@@ -73,6 +76,7 @@ func Start(ctx context.Context, httpPort, httpsPort int, scaledContext *config.S
 	root.PathPrefix("/k8s/clusters/").Handler(authedHandler)
 	root.PathPrefix("/meta").Handler(authedHandler)
 	root.NotFoundHandler = ui.UI(http.NotFoundHandler())
+	root.PathPrefix("/v1-saml/").Handler(samlRoot)
```
Comment (cjellick) [inline server/server.go]: most of the other PathPrefix functions do not have a trailing /. I think yours should not have it either. The endpoint "/v1-saml" is technically your httpHandler's responsibility to handle (even if it just 404s).

File: pkg/auth/providers/saml/saml_client.go [hunk 14]
```diff
@@ -0,0 +1,277 @@
...
+	config *v3.SamlConfig
+	SamlSP *samlsp.Middleware
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: this variable is not used. can we remove it?

File: pkg/auth/providers/saml/saml_client.go [hunk 15]
```diff
@@ -0,0 +1,277 @@
...
+}
+
+const (
+	redirectBackBase = "redirectBackBase"
+	redirectBackPath = "redirectBackPath"
+)
+
+var Root *mux.Router
+
+func (s *Provider) InitializeSPClient(configToSet *v3.SamlConfig, name string) error {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: As mentioned elsewhere, this should not be attached to any struct.
Also, can you change its name? it is not initializing a client.
Can you also move it to a different file?

File: pkg/auth/providers/saml/saml_client.go [hunk 16]
```diff
@@ -0,0 +1,277 @@
...
+		log.Errorf("SAML: HandleSamlPost failed to parse query params with error %v", err)
+		w.WriteHeader(500)
+		w.Write([]byte("SAML: Failed to get auth query parameters"))
+		return
+	}
+
+	// SAML library middleware, on a request with valid session, returns the assertion attributes within request's context
+	samlData := make(map[string][]string)
+	samlData = samlsp.Token(r.Context()).Attributes
+	log.Infof("SAML: HandleSamlPost: Received a SAML POST data %v", samlData)
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: probably should be at debug, not info

File: pkg/auth/providers/saml/saml_provider.go [hunk 17]
```diff
@@ -0,0 +1,292 @@
...
+	if config == nil {
+		config, err = s.getSamlConfig()
+		if err != nil {
+			return v3.Principal{}, nil, nil, err
+		}
+	}
+	s.SamlClient.config = config
+	inputMap := samlCredential.Code
+
+	if err := json.Unmarshal([]byte(inputMap), &samlData); err != nil {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: Where to we verify/validate this token?

File: pkg/auth/providers/saml/saml_provider.go [hunk 18]
```diff
@@ -0,0 +1,292 @@
...
+	var samlData map[string][]string
+
+	if config == nil {
+		config, err = s.getSamlConfig()
+		if err != nil {
+			return v3.Principal{}, nil, nil, err
+		}
+	}
+	s.SamlClient.config = config
+	inputMap := samlCredential.Code
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: Where do we verify/validate this code/token?

File: pkg/auth/providers/publicapi/login.go [hunk 19]
```diff
@@ -130,9 +138,17 @@ func (h *loginHandler) createLoginToken(request *types.APIContext) (v3.Token, st
 	}
 
 	// Authenticate User
-	userPrincipal, groupPrincipals, providerToken, err := providers.AuthenticateUser(input, providerName)
-	if err != nil {
-		return v3.Token{}, "", err
+	if providerName != saml.PingName {
```
Comment (cjellick) [inline pkg/auth/providers/publicapi/login.go]: I want a more clearly defined boundary between saml and non-saml flow. THis doesnt make it obvious why saml.PingName is special. Can you rearrrange to something like:
```
// Some comment exlpaining that saml has a different flow that requires us to short circuit the normal flow
if providerName == saml.Ping {
   ...
}

userPrincipal, groupPrincipals, providerToken, err = providers.AuthenticateUser(input, providerName)
```

File: pkg/auth/providers/publicapi/login.go [hunk 20]
```diff
@@ -130,9 +138,17 @@ func (h *loginHandler) createLoginToken(request *types.APIContext) (v3.Token, st
...
+		userPrincipal, groupPrincipals, providerToken, err = providers.AuthenticateUser(input, providerName)
+		if err != nil {
+			return v3.Token{}, "", err
+		}
+	} else {
+		err := saml.PerformAuthRedirect(providerName, request)
+		if err != nil {
+			return v3.Token{}, "saml", nil
+		}
+		return v3.Token{}, "saml", nil
```
Comment (cjellick) [inline pkg/auth/providers/publicapi/login.go]: Whether err is nil or not, you do the same thing; you can drop the err check

File: pkg/auth/providers/saml/saml_account.go [hunk 21]
```diff
@@ -0,0 +1,8 @@
+package saml
+
+type Account struct {
+	UID         string `json:"uid,omitempty"`         //DN
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_account.go]: Why do we call this field UID? Does it have special meaning in the saml world? Can we just call it it ID or PrincipalId?
For that matter, do we even need this Account struct or can we go straight to the Principal struct? You may have been alluding to this yourself the other day.

File: pkg/auth/providers/saml/saml_actions.go [hunk 22]
```diff
@@ -82,55 +85,18 @@ func (s *Provider) testAndApply(actionName string, action *types.Action, request
 	}
 
 	samlConfig = samlConfigApplyInput.SamlConfig
-	s.Request = request
+	p := SamlProviders[s.Name]
+	p.SamlClient.config = &samlConfig
+	p.Request = request
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_actions.go]: Discussed offline, but making a note here so that we do not forget it: this is sharing state across api requests. Wont work in HA scenarios and potentially others (like if a user abandons the flow),

File: pkg/api/controllers/samlconfig/samlconfig.go [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+	samlConfig := &v3.SamlConfig{}
+	if key == "" || config == nil {
+		return nil
+	}
+
+	if config.Name != saml.PingName {
+		return nil
+	}
+
+	if config.Annotations["configured"] != "true" {
```
Comment (cjellick) [inline pkg/api/controllers/samlconfig/samlconfig.go]: I think this annotation is going to cause issues, but im not sure what yet because I havent fully reviewed the PR.
one issue: it is never set to false. So, if the provider is disabled, this annotation will still be true.

But im not even sure if we need it anymore. I believe it was originally needed when we had the function broken into two: configureTest and testAndApply

File: pkg/auth/providers/saml/saml_provider.go [hunk 24]
```diff
@@ -0,0 +1,237 @@
...
+	"github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const PingName = "ping"
+
+type Provider struct {
+	Ctx             context.Context
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: prooperties on a struct should only be exported if they are used outside the package

File: pkg/auth/providers/saml/saml_provider.go [hunk 25]
```diff
@@ -0,0 +1,237 @@
...
+}
+
+func (s *Provider) CustomizeSchema(schema *types.Schema) {
+	schema.ActionHandler = s.actionHandler
+	schema.Formatter = s.formatter
+}
+
+func (s *Provider) TransformToAuthProvider(authConfig map[string]interface{}) map[string]interface{} {
+	p := common.TransformToAuthProvider(authConfig)
+	if s.Name == PingName {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: Is the logic going to vary from provider to provider? seems like it should be the same logic and so we should not need if blocks for each provider. If the only difference is the name of the constant we use, I think we can just safely use the same constant for all providers

File: pkg/auth/providers/saml/saml_provider.go [hunk 26]
```diff
@@ -0,0 +1,237 @@
...
+		p[publicclient.PingProviderFieldRedirectURL] = formSamlRedirectURLFromMap(authConfig, s.Name)
+	}
+	return p
+}
+
+func (s *Provider) AuthenticateUser(input interface{}) (v3.Principal, []v3.Principal, string, error) {
+	return v3.Principal{}, nil, "", nil
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_provider.go]: add a comment explaining why this is not implemented

File: pkg/api/controllers/samlconfig/samlconfig.go [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func newAuthProvider(apiContext *config.ScaledContext) *authProvider {
+	a := &authProvider{
+		authConfigs: apiContext.Management.AuthConfigs(""),
+	}
+	return a
+}
+
+func (a *authProvider) sync(key string, config *v3.AuthConfig) error {
```
Comment (cjellick) [inline pkg/api/controllers/samlconfig/samlconfig.go]: This sync method is potentially called a lot, even when there has not been a change to the config. We should add some intelligene here or in `InitializeSamlServiceProvider`so that we are smart about when we reinitialize

File: pkg/auth/providers/saml/saml_provider.go (hunk 26 above)
Comment (cjellick) [reply pkg/auth/providers/saml/saml_provider.go]: in fact, lets return an error saying that this method is not supported

File: pkg/auth/providers/saml/saml_client.go [hunk 28]
```diff
@@ -0,0 +1,385 @@
...
+	var privKey *rsa.PrivateKey
+	var cert *x509.Certificate
+	var err error
+	var ok bool
+
+	if configToSet.IDPMetadataURL == "" {
+		idpURL = ""
+		if configToSet.IDPMetadataContent == "" {
+			if configToSet.IDPMetadataFilePath == "" {
+				log.Debugf("SAML: Cannot initialize saml SP properly, missing IDP URL/metadata in the config %v", configToSet)
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: should we return an error here?

Comment (cjellick) [reply pkg/auth/providers/saml/saml_client.go]: same for lines 73 and/or 78

File: pkg/auth/providers/saml/saml_client.go [hunk 29]
```diff
@@ -0,0 +1,385 @@
...
+			}
+		default:
+			return fmt.Errorf("SAML: unsupported key type %q", block.Type)
+		}
+	}
+
+	if configToSet.SPSelfSignedCert != "" {
+		block, _ := pem.Decode([]byte(configToSet.SPSelfSignedCert))
+		if block == nil {
+			panic("SAML: failed to parse PEM block containing the private key")
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: no panics in the code. return errors and ensure it is handled properly higher up. same on line 133

File: pkg/auth/providers/saml/saml_client.go [hunk 30]
```diff
@@ -0,0 +1,385 @@
...
+
+	if name == PingName {
+		Root.Get("PingLogin").HandlerFunc(provider.HandleSamlLogin)
+		Root.Get("PingACS").HandlerFunc(provider.ServeHTTP)
+		Root.Get("PingMetadata").HandlerFunc(provider.ServeHTTP)
+	}
+	return nil
+}
+
+func PingHandlers() *mux.Router {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: this is called from outside of the auth packages. caller shouldnt need to know its for ping specifically.
rename as `SAMLAuthHandler` and change the return type to just http.Handler. Caller should not know implementation details of what is being returned

File: pkg/auth/providers/saml/saml_client.go [hunk 31]
```diff
@@ -0,0 +1,385 @@
...
+	IDPSSODescriptors []saml.IDPSSODescriptor `xml:"IDPSSODescriptor"`
+	SPSSODescriptors  []saml.SPSSODescriptor  `xml:"SPSSODescriptor"`
+}
+
+const (
+	redirectBackBase = "redirectBackBase"
+	redirectBackPath = "redirectBackPath"
+)
+
+var Root *mux.Router
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: should not be exported

File: pkg/auth/providers/saml/saml_handlers.go [hunk 32]
```diff
@@ -0,0 +1,126 @@
...
+		w.Write(buf)
+		return
+	}
+
+	if r.URL.Path == serviceProvider.AcsURL.Path {
+		r.ParseForm()
+		assertion, err := serviceProvider.ParseResponse(r, s.getPossibleRequestIDs(r))
+		if err != nil {
+			if parseErr, ok := err.(*saml.InvalidResponseError); ok {
+				serviceProvider.Logger.Printf("RESPONSE: ===\n%s\n===\nNOW: %s\nERROR: %s",
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_handlers.go]: just use our normal logging paradigm

File: pkg/auth/providers/saml/saml_actions.go [hunk 33]
```diff
@@ -0,0 +1,96 @@
...
+	}
+
+	data := map[string]interface{}{
+		"redirectUrl": redirectURL,
+		"type":        "samlConfigTestOutput",
+	}
+
+	annotations := make(map[string]string)
+	annotations["configured"] = "true"
+	annotations["userID"] = s.UserMGR.GetUser(request)
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_actions.go]: saving the userID as an annotation is bad idea. we need to find a better way

Comment (cjellick) [reply pkg/auth/providers/saml/saml_actions.go]: set the userID as a claim on the signed jwt cookie that we are using for "state"

File: pkg/auth/providers/saml/saml_client.go [hunk 34]
```diff
@@ -0,0 +1,385 @@
...
+
+	if !config.Enabled {
+		header := make(map[string][]string)
+		header["Impersonate-User"] = []string{config.Annotations["userID"]}
+		newReq := &types.APIContext{
+			Request: &http.Request{
+				Header: header,
+			},
+		}
+		user, err := s.UserMGR.SetPrincipalOnCurrentUser(newReq, userPrincipal)
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: instead of making the fake request and passing it in, can you expose a new function on userMGR that takes in a userID instead of an entire request?

File: pkg/auth/providers/saml/saml_client.go [hunk 35]
```diff
@@ -0,0 +1,385 @@
...
+	if err != nil {
+		writeError(w, 500, "SAML: Error during login while checking access %v", err, "Server error while authenticating")
+		return
+	}
+	if !allowed {
+		writeError(w, 403, "SAML: User does not have access %v", err, "User does not have access")
+		return
+	}
+
+	if !config.Enabled {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: as discussed, make this decision based off of claims in the jwt, not the state of enabled

File: pkg/auth/providers/saml/saml_client.go [hunk 36]
```diff
@@ -0,0 +1,385 @@
...
+			},
+		}
+		user, err := s.UserMGR.SetPrincipalOnCurrentUser(newReq, userPrincipal)
+		if err != nil {
+			writeError(w, 500, "SAML: Error setting principal on current user %v", err, "Server error while authenticating")
+			return
+		}
+
+		config.Enabled = true
+		err = s.saveSamlConfig(config)
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: dont change the config or save it. just rely on the user having explicitly enabled it

File: pkg/auth/providers/saml/saml_client.go (hunk 35 above)
Comment (cjellick) [reply pkg/auth/providers/saml/saml_client.go]: when you change this logic, please reorganize to make it very clear that one code path is for the "testing" scenario and the other is for normal login. Like, even make different functions for each of those code paths for this function to call

File: pkg/auth/providers/saml/saml_client.go [hunk 37]
```diff
@@ -0,0 +1,385 @@
...
+	}
+
+	setRancherToken(w, r, s.TokenMGR, user.Name, userPrincipal, groupPrincipals, true)
+	return
+}
+
+func setRancherToken(w http.ResponseWriter, r *http.Request, tokenMGR *tokens.Manager, userID string, userPrincipal v3.Principal,
+	groupPrincipals []v3.Principal, isSecure bool) {
+	rToken, err := tokenMGR.NewLoginToken(userID, userPrincipal, groupPrincipals, "", 0, "")
+	if err != nil {
```
Comment (cjellick) [inline pkg/auth/providers/saml/saml_client.go]: dont have setRancherToken writeError. Have it return an error and the caller decide what to do with it (write it)

File: pkg/auth/providers/saml/saml_provider.go (hunk 25 above)
Comment (mrajashree) [reply pkg/auth/providers/saml/saml_provider.go]: We're also using the redirectURL from different providers, Ping vs Shibboleth, hence the if-else

Comment (cjellick) [reply pkg/auth/providers/saml/saml_provider.go]: ok, cool

**Comment by cjellick** [comment]
LGTM
